### PR TITLE
refactor: Activate/deactivate function to handle events and publish to Firestore

### PR DESCRIPTION
**Pull request by danny-molnar**

### CODE REVIEW

File: functions/activate_deactivate/entrypoint.go [hunk 1]
```diff
@@ -32,31 +34,98 @@ func init() {
...
+	if err != nil {
+		return fmt.Errorf("(ACG Client) Error creating client: %s", err)
+	}
+
+	var action string = activateMessage.RequestType // can be "activate", "deactivate", "invite"
 
-	/*	clientAcg, err := acg.NewClient(&serviceConfig.ApiKey, &serviceConfig.ConsumerId)
+	// Perform the requested action
+	var actionResponse *acg.Response
+	if action == "invite" || activateMessage.UserID == "" {
```
Comment (riweston) [inline functions/activate_deactivate/entrypoint.go:66]: - I would restructure this slightly, pop this into a case/switch statement so it's clear what the eventualities are and they're all in one place if you have multiple cases though i'm not sure you do accounting for my next point
- `invite` implicitly wouldn't have a `UserID`, hence why the action is `invite`

File: functions/activate_deactivate/entrypoint.go [hunk 2]
```diff
@@ -32,31 +34,98 @@ func init() {
...
-		//clientAcg.SetUserActivated(&(userAcg), false)
-		slackMsg = fmt.Sprintln("User", userAcg.UserId, "has been deactivated by admin")
+	if err != nil {
+		return fmt.Errorf("failed to process '%s' action: %v", action, err)
 	}
 
-	clientSlack := slack.New(serviceConfig.SlackBotToken)
-	clientSlack.PostMessage("", slack.MsgOptionReplaceOriginal(request.ResponseUrl), slack.MsgOptionText(slackMsg, false))
+	// Log action to Firestore
+	docID := fmt.Sprintf("%s-%s", action, activateMessage.UserID)
```
Comment (riweston) [inline functions/activate_deactivate/entrypoint.go:92]: Can we rely on an auto key here? Multiple actions on the same user would result in the firestore doc getting overwritten so we would lose history over time

**Review by riweston — APPROVED** [review]

**Review by riweston — CHANGES_REQUESTED** [review]
Misclick before, a few things to review

File: functions/activate_deactivate/entrypoint.go (hunk 2 above)
Comment (danny-molnar) [reply functions/activate_deactivate/entrypoint.go:92]: Yes, that makes total sense...

File: functions/activate_deactivate/entrypoint.go [hunk 3]
```diff
@@ -32,31 +34,98 @@ func init() {
...
+			"timestamp": time.Now(),
+			"source":    activateMessage.Source,
+		},
+	}
+	if err := firestoreClient.PublishDocument(ctx, doc, fmt.Sprintf("%sEvents", action)); err != nil {
+		return fmt.Errorf("failed to publish %s event to Firestore: %w", action, err)
+	}
+
+	// Send a message to Slack if a response URL is provided
+	if activateMessage.ResponseUrl != "" {
```
[the remark this answers is not in the document]
Comment (danny-molnar) [reply functions/activate_deactivate/entrypoint.go:111]: Yep, best to be clear.

File: functions/activate_deactivate/entrypoint.go (hunk 1 above)
Comment (danny-molnar) [reply functions/activate_deactivate/entrypoint.go:66]: Right, let me refactor that logic there.

**Review by riweston — APPROVED** [review]

**Pull request merged**